### ISSUE TITLE
questions about qgen_lstm_network file

### ISSUE THREAD

**Issue by JasZhanAva** [GuessWhatGame/guesswhat#16]
Hi,

I'm reading the qgen_lstm_network.py file, and really confused about line 121 and line 152, that ml_loss and log_of_policy take the identified form of self.cross_entropy_loss ?

From my understanding, the self.cross_entropy_loss is a type of loss? How can log_of_policy just takes tf.identity(self.cross_entropy_loss)? Really appreciate some interpretations. Thanks!

**Comment by harm-devries** [comment]
I agree these lines of code are confusing. One thing you'll need to realize is that for a one-hot target distribution y, the ```cross_entropy(prob, y) = - log prob[i]``` where i is the non-zero index of y. 

This term also pops up in the reinforce update ```log prob[i] * r ``` where i denotes the sampled action. So if you set the target distribution y to the sampled actions, you obtain the correct update if you multiply it with the reward (and correct for the minus). 

This is how it's implemented in this code base, but I do agree that it would be easier to understand if it was implemented with a tf.log and tf.gather over the softmax_output.

**Comment by JasZhanAva** [comment]
@harmdevries89 Thanks for the explanation. :) 
I have another question regarding line140, the flat_baseline dimension is supposed to be batch_size*1, why it can be reshaped as line140?

**Comment by harm-devries** [comment]
We predict the state-value baseline each time we sample an action i.e. at each timestep of the decoder. That's why the variable ```decoder_out```, which is of dimension ```[batch_size, seq_len, lstm_cell_size]```, needs to be reshaped to ```[batch_size*seq_len, lstm_cell_size]``` before it's passed to the MLP (on line 141). At line 144, it's then reshaped back to a 3-tensor.

**Comment by JasZhanAva** [comment]
Got it, thanks! :)
